### PR TITLE
blockchain: Overhaul and update to sync to btcd.

### PR DESCRIPTION
**Pull request by cjepson**
Before merging, the following components will be synced in from btcd:

- [x] Database architectural overhaul
- [x] Integration of the new database and updating of the underlying consensus code
- [x] Make the database more easily upgradeable by tracking version of the database and the script compression algorithm
- [x] Integration of the new indexers for transactions, the address index, and a new address index tracking ever used addresses
- [x] Removal of the old database and replacing it with the new database everywhere in the codebase

After this is complete, work can begin on overhauling the ticket memory database so that instead uses the new ticket treap.

### CODE REVIEW

**Comment by jcvernaleo** [comment]
I've been trying to test this on one of my testnet nodes (bitrig, i3, slow spinning disk).
IBD took a few hours (not bad for that machine).  Since I didn't have my wallet on the correct commit I went back to master and it took about 12 minutes to resync the ticket db.
Got my wallet correct the next day and I restarted daemon on this branch.  So the block db is about 1 day behind and it needed to resync the ticket db (after I deleted the old file).  It has been on the line:
```
13:53:15 2016-07-13 [INF] STKE: Db non-empty, resyncing ticket DB
```
for the last 18 hours.  That is longer than an IDB would take so I tend to think something is wrong.

File: blockchain/subsidy.go [hunk 1]
```diff
@@ -26,36 +64,63 @@ import (
...
+	iteration := uint64(height / s.params.ReductionInterval)
+
+	if iteration == 0 {
+		return s.params.BaseSubsidy
+	}
+
+	// First, check the cache.
+	s.subsidyCacheLock.RLock()
+	if cachedValue, existsInCache := s.subsidyCache[iteration]; existsInCache {
+		subsidy := cachedValue
```
Comment (davecgh) [inline blockchain/subsidy.go]: `cachedValue` is already a copy, so there is no need to copy it again.  You can just delete this and do:

```Go
	s.subsidyCacheLock.RUnlock()
	return cachedValue
```

Comment (davecgh) [inline blockchain/subsidy.go]: Actually, I'd just make this consistent with the code below by only doing the unlock once around the assignments:

```Go
s.subsidyCacheLock.RLock()
cachedValue, existsInCache := s.subsidyCache[iteration];
s.subsidyCacheLock.RUnlock()
if existsInCache {
	return cachedValue
}
```

File: blockchain/subsidy.go [hunk 2]
```diff
@@ -26,36 +64,63 @@ import (
...
+		s.subsidyCacheLock.RUnlock()
+		return subsidy
+	}
+	s.subsidyCacheLock.RUnlock()
+
+	// Is the previous one in the cache? If so, calculate
+	// the subsidy from the previous known value and store
+	// it in the database and the cache.
+	s.subsidyCacheLock.RLock()
+	cacheValue, existsInCache := s.subsidyCache[iteration-1]
```
Comment (davecgh) [inline blockchain/subsidy.go]: Then you can just make this `=` instead of `:=`.

Comment (cjepson) [inline blockchain/subsidy.go]: Thanks, updating now with this. I also forgot to cache the fully computed value at the end, so now that is also fixed.

**Comment by jolan** [comment]
Mining against this fine.

tACK

**Comment by jcvernaleo** [comment]
Thanks to the flaky machine I run a testnet node on, I was able to test how this deals with an unclean shutdown and unlike the early version, it seems to have recovered just fine now.

**Comment by jrick** [comment]
I'm in favor of merging this in now so long as the associated dcrrpcclient fix goes in as well so dcrctl remains usable.  Too many things blocked on this reaching master and no blocking errors have been discovered from it yet.

**Comment by marcopeereboom** [comment]
I want to move forward on this.  As long as this sits on a branch it'll not get enough attention and get tested properly.

**Comment by alexlyp** [comment]
have tested extensively and everything seems stable, but have not been able to review all code.  Since much of it was extensively reviewed upstream in btcd, I am very confident in it's quality.  We should be able to shake out any further issues before it is packaged for release.

**Pull request merged**